### ISSUE TITLE
Daily schedules not allowing to select the entire day

### ISSUE THREAD

**Issue by maciejdlugosz** [BOINC/boinc#2583]
In Options->Computing preferences->Daily schedules it is impossible to set a day, when computation or transfering are not limited. E.g. we would like to compute on every day only at night, but on weekend the entire day. Before #2044 it was possible to pass a range 0:00 to 24:00.

Passing 0:00 to 23:59 is not a huge cost until one has "Leave non-GPU tasks in memory while suspended" checked.

I suggest not to add a new checkbox, but rather allow passing "0:00 to 24:00" or more intuitively "0:00 to 0:00".

**Comment by AenBleidd** [comment]
@CharlieFenton, 
why did this change was done 37ae3de4 ?
It looks like that next code accepts equal start and end time without any problems:

```
bool TIME_SPAN::suspended(double hour) const {
    if (start_hour == end_hour) return false;
    if (start_hour == 0 && end_hour == 24) return false;
    if (start_hour == 24 && end_hour == 0) return true;
    if (start_hour < end_hour) {
        return (hour < start_hour || hour > end_hour);
    } else {
        return (hour >= end_hour && hour < start_hour);
    }
}

TIME_SPAN::TimeMode TIME_SPAN::mode() const {
    if (end_hour == start_hour || (start_hour == 0 && end_hour == 24)) {
        return Always;
    } else if (start_hour == 24 && end_hour == 0) {
        return Never;
    }
    return Between;
}
```

**Comment by CharlieFenton** [comment]
@AenBleidd 

> why did this change was done 37ae3de ?

Change 37ae3de has nothing to do with this PR. Even though _GIT blame_ lists me as the author of that code, commit 98cfb8d on 26 September 2008 lists @davidpanderson as the author of this code. But please note that the code you reference above is is not called by the Manager's dialog. I also don't see anywhere in the client or Manager where this code is called. 

As stated in [the original complaint](https://github.com/BOINC/boinc/issues/2583#issue-337308547), the issue was caused by your PR  #2044 (and related PR #2083) and your commits f39005c, 7b8cfd6 and 7b8cfd6. I believe you should change this:
```
    //verify hours
    if (h < 0 || h > 23) {
        return false;
    }

```
to this:
```
    //verify hours
    if (h < 0 || h > 24) {
        return false;
    }

```
The 24 case should be handled by this code (to ensure that it is _24:00_:
```
    //all chars are valid, now what is with the value as a whole ?
    wxDateTime dt;
    const wxChar* stopChar = dt.ParseFormat(value,wxT("%H:%M"));
    if(stopChar==NULL && value != wxT("24:00")) {
        // conversion failed
        return false;
    }

```

but I have not tested it throughly.

**Comment by CharlieFenton** [comment]
@AenBleidd and @davidpanderson The default values set in the Computing Preferences Dialog (and the global preferences) by GLOBAL_PREFS::enabled_defaults() is 23:59:
```
    cpu_times.end_hour = 23.983;    // 23:59
    net_times.start_hour = 0;
    net_times.end_hour = 23.983;    // 23:59

```
Since we are allowing 0:00 to 24:00, should this be 24.0 instead? Please note the comment on this code:
`// These should be consistent with html/inc/prefs.inc`

**Comment by CharlieFenton** [comment]
@AenBleidd,
> The function TIME_SPAN::suspended is used in client as far as I see.

Yes, I see it now. It is called as `global_prefs.cpu_times.suspended(...)` and `global_prefs.cpu_times.suspended(...)`

> It is incorrect just to use this:
> 
>     //verify hours
>     if (h < 0 || h > 24) {
>         return false;
>     }
> 
> Because in this case user can enter 24:59 that is not correct.
> 

As I stated in my earlier comment, that is prevented by this line in _CDlgAdvPreferences::IsValidTimeValue()_, which disallows any value between 24:01 and 24:59:
`    if(stopChar==NULL && value != wxT("24:00")) {
`
I tested this briefly and it worked correctly for me. 

On looking up the description of wxDateTime::ParseFormat() on the wxWidget website I see that this form of the function is no longer listed. But the date time.h header file from wxWidgets 3.1.0 says:
```
    // backwards compatible versions of the parsing functions: they return an
    // object representing the next character following the date specification
    // (i.e. the one where the scan had to stop) or a special NULL-like object
    // on failure
    //
    // they're not deprecated because a lot of existing code uses them and
    // there is no particular harm in keeping them but you should still prefer
    // the versions above in the new code
``` 

> I can fix this verification but I just curious why the changes in the commit 37ae3de were done?

When I look at that link, GITHub shows me code for the web preferences interface that makes no mention of 23:59, and the branch to which it refers is about a fix to the _Notices_ tab and has nothing to do with the Preferences dialog:
`client_release/7.6/OSX_notices_hotfix/7.6.34`

The commit matching this which I find in the GIT log is 3148f84. This was added because there were no initial values being set when a user checked the box in the dialog. Sorry, I do not remember why we used those values. I probably thought that setting the times from 0:00 to 23:59 would enable computing or network for the entire day. But looking at the code now, I see that it would suspend for the last minute of the day, because the times in `TIME_SPAN::suspended()` are floating point values so fractions of seconds matter.

**Comment by CharlieFenton** [comment]
@AenBleidd,
> The function TIME_SPAN::suspended is used in client as far as I see.

Yes, I see it now. It is called as `global_prefs.cpu_times.suspended(...)` and `global_prefs.cpu_times.suspended(...)`

> It is incorrect just to use this:
> 
>     //verify hours
>     if (h < 0 || h > 24) {
>         return false;
>     }
> 
> Because in this case user can enter 24:59 that is not correct.
> 

As I stated in my earlier comment, that is prevented by this line in _CDlgAdvPreferences::IsValidTimeValue()_, which disallows any value between 24:01 and 24:59:
`    if(stopChar==NULL && value != wxT("24:00")) {
`
I tested this briefly and it worked correctly for me. 

On looking up the description of wxDateTime::ParseFormat() on the wxWidget website I see that this form of the function is no longer listed. But the date time.h header file from wxWidgets 3.1.0 says:
```
    // backwards compatible versions of the parsing functions: they return an
    // object representing the next character following the date specification
    // (i.e. the one where the scan had to stop) or a special NULL-like object
    // on failure
    //
    // they're not deprecated because a lot of existing code uses them and
    // there is no particular harm in keeping them but you should still prefer
    // the versions above in the new code
``` 

> I can fix this verification but I just curious why the changes in the commit 37ae3de were done?

When I look at that link, GITHub shows me code for the web preferences interface that makes no mention of 23:59, and the branch to which it refers is about a fix to the _Notices_ tab and has nothing to do with the Preferences dialog:
`client_release/7.6/OSX_notices_hotfix/7.6.34`

The commit matching this which I find in the GIT log is 3148f84. This was added because there were no initial values being set when a user checked the box in the dialog. Sorry, I do not remember why we used those values. I probably thought that setting the times from 0:00 to 23:59 would enable computing or network for the entire day. But looking at the code now, I see that it would disallow it for the last minute of the day, because the times in `TIME_SPAN::suspended()` are floating point values so fractions of seconds matter.

One more thought: instead of the somewhat convoluted code testing now for 24:00, it might be 
better to just add this:
`    if ((h==24) && (m==0)) return true;
`
before this:
 ```
   //verify hours
    if (h < 0 || h > 23) {
        return false;
    }
```
 leaving the `h > 23` unchanged
and eliminate all this:
```
    //all chars are valid, now what is with the value as a whole ?
    wxDateTime dt;
    const wxChar* stopChar = dt.ParseFormat(value,wxT("%H:%M"));
    if(stopChar==NULL && value != wxT("24:00")) {
        // conversion failed
        return false;
    }
```

**Comment by CharlieFenton** [comment]
@AenBleidd,
I did a bit more testing. The code you have now still allows bad input of the form `2:3:4`. The following code will fix that and will also fix the issue raised in PR #2044:
```
/* checks if the value is a valid time */
bool CDlgAdvPreferences::IsValidTimeValue(const wxString& value) {
    for(unsigned int i=0; i < value.Length();i++) {
        if(!IsValidTimeChar(value[i])) {
            return false;
        }
    }
    //verify correct format and range of time values
    int h = -1, m = -1;
    //verify the format itself
    int parsed = sscanf(value.c_str(), "%d:%d", &h, &m);
    if (parsed != 2) {
        return false;
    }
     //all chars are valid, now what is with the value as a whole ?
    wxDateTime dt;
    const wxChar* stopChar = dt.ParseFormat(value,wxT("%H:%M"));
    if(stopChar==NULL && value != wxT("24:00")) {
        // conversion failed
        return false;
    }
    if(stopChar!='\0') return false;
    return true;
}
```
In the case of `2:3:4` _stopChar_ will be ':' for `13:239` _stopChar_ will be '9' and for `18:555` _stopChar_ will be '5'.

**Comment by CharlieFenton** [comment]
I neglected to mention that if the time values are outside the normal range (0 to 23 for hours and 0 to 59 for minutes) _dt.ParseFormat()_ will set _stopChar_ to NULL, so `if(stopChar!='\0') return false;` will return _false_, so there is no need to separately test for that.

**Comment by CharlieFenton** [comment]
@AenBleidd,
I finally had a chance to actually test and debug my suggested code. Here is the correct, tested code:
```
/* checks if the value is a valid time */
bool CDlgAdvPreferences::IsValidTimeValue(const wxString& value) {
    for(unsigned int i=0; i < value.Length();i++) {
        if(!IsValidTimeChar(value[i])) {
            return false;
        }
    }
    //verify correct format and range of time values
    int h = -1, m = -1;
    //verify the format itself
    int parsed = sscanf(value.c_str(), "%d:%d", &h, &m);
    if (parsed != 2) {
        return false;
    }
    if (value == wxT("24:00")) return true;
    //all chars are valid, now what is with the value as a whole ?
    wxDateTime dt;
    const wxChar* stopChar = dt.ParseFormat(value,wxT("%H:%M"));
    if(stopChar==NULL) return false;    // conversion failed
    if(*stopChar!='\0') return false;   // conversion failed
    return true;
}
```

**Comment by AenBleidd** [comment]
Thank you, @CharlieFenton, I'll review and test this a little bit later, hopefully within next two days.

**Comment by CharlieFenton** [comment]
@AenBleidd,
Actually, I'm pretty sure that the _fscanf()_ is not needed. The following code should be sufficient:
```
/* checks if the value is a valid time */
bool CDlgAdvPreferences::IsValidTimeValue(const wxString& value) {
    for(unsigned int i=0; i < value.Length();i++) {
        if(!IsValidTimeChar(value[i])) {
            return false;
        }
    }
    //all chars are valid, now what is with the value as a whole ?
    if (value == wxT("24:00")) return true;
    wxDateTime dt;
    const wxChar* stopChar = dt.ParseFormat(value,wxT("%H:%M"));
    if(stopChar==NULL) return false;    // conversion failed
    if(*stopChar!='\0') return false;   // conversion failed
    return true;
}
```

**Comment by AenBleidd** [comment]
Pull request is created. Please review

**Comment by AenBleidd** [comment]
Fixed via #2589 and #2591.
